Sources/MessagingPush/PushHandling/PushEventHandlerProxy.swift

```diff
@@ -39,8 +39,18 @@ class PushEventHandlerProxyImpl: PushEventHandlerProxy {
             return
         }
 
-        nestedDelegates.forEach { _, delegate in
-            delegate.onPushAction(pushAction, completionHandler: completionHandler)
+        Task {
+            // Wait for all other push event handlers to finish before calling the completion handler.
+            // Each iteration of the loop waits for the push event to be processed by the delegate.
+            for delegate in nestedDelegates.values {
```

Comment:
great start !! but i do believe we are still prone to race condition, because we are not explicitly handling concurrent access to the `nestedDelegates` dictionary, which could lead to race conditions if multiple threads attempted to modify this shared resource simultaneously.

what do you think about making `nestedDelegates` as an actor?

Reply:
I changed `nestedDelegates` to be `@Atomic` which should solve the problem you mentioned. 

I do see a lot of value in `actor`. For this situation, I think it would require a bigger refactor to use it. 

To use `actor` to protect `nestedDelegates`, we would need to make the Proxy class an `actor` instead of a `class`. If we made that change, then we would also need to make functions such as `addPushEventHandler` to be `async`. That change would require a lot of code changes in the project that I do not think is worth it at this time.

Reply:
Before guard condition add something like `let nestedDelegatesCopy = nestedDelegates`
 Then use the `nestedDelegatesCopy` throughout the method. This ensures the array copy you checked at the beginning of the method, won't change in the middle of the method as it is a local copy.

Reply:
[Made change](https://github.com/customerio/customerio-ios/compare/97e5e022765adf66f9baf088b160173d9bf284e5..e8c670f6722df673959cd887a98002a1f4f42f81)

Reply:
I'll be okay with these changes for now. 

Regarding:
> To use an actor to protect nestedDelegates, we would need to make the Proxy class an actor instead of a class. If we made that change, then we would also need to make functions such as addPushEventHandler to be async

I don't think we need to do all that, we can just do something like this and it should work too. 

```
actor NestedDelegatesActor {
    private var nestedDelegates: [String: PushEventHandler] = [:]

    var hasDelegates: Bool {
        !nestedDelegates.isEmpty
    }

    func addDelegate(_ delegate: PushEventHandler) {
        nestedDelegates[String(describing: delegate)] = delegate
    }

    func removeDelegate(_ delegate: PushEventHandler) {
        nestedDelegates.removeValue(forKey: String(describing: delegate))
    }

    func forEachDelegate(_ action: @escaping (PushEventHandler) async -> Void) async {
        for delegate in nestedDelegates.values {
            await action(delegate)
        }
    }
}
```

```
class PushEventHandlerProxyImpl: PushEventHandlerProxy {
    /*
     # Why is this class not stored in the digraph?

     Similar to why the SDK's `UNUserNotificationCenterDelegate` instance is also not in the digraph. See those comments to learn more.
     */
    public static let shared = PushEventHandlerProxyImpl()

    // Use a map so that we only save 1 instance of a given handler.
    private var nestedDelegates = NestedDelegatesActor()

    func addPushEventHandler(_ newHandler: PushEventHandler) {
        Task {
            await nestedDelegates.addDelegate(newHandler)
        }
    }

    func onPushAction(_ pushAction: PushNotificationAction, completionHandler: @escaping () -> Void) {
        Task {
            await nestedDelegates.forEachDelegate { delegate in
                await withCheckedContinuation { continuation in
                    delegate.onPushAction(pushAction) {
                        continuation.resume()
                    }
                }
            }
            completionHandler()
        }
    }
 }
```

but i am okay with revisiting it later.